### PR TITLE
Make Localstack services optional.

### LINKED ISSUE #6428: [Enhancement]: Localstack services should be optional
### Module

LocalStack

### Proposal

Starting from [Localstack 0.13.0](https://github.com/localstack/localstack/releases/tag/v0.13.0) the services are loaded lazily and there is no need to list them when starting Localstack container. Thus `LocalstackContainer` should not fail to start if the services list are not provided.

There probably should be still an option to set services for eagerly loaded services.

### PR DESCRIPTION
**Pull request by maciejwalkowiak**
Fixes #6428

### CODE REVIEW

File: modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java [hunk 1]
```diff
@@ -61,6 +61,15 @@ public class LocalStackContainer extends GenericContainer<LocalStackContainer> {
...
     private final boolean legacyMode;
 
+    /**
+     * Starting with version 0.13.0, setting services list on Localstack is not required. When <code>false</code>,
+     * containers are started lazily. When <code>true</code>, container fails to start if services list is not provided.
+     *
+     * Testcontainers will use the tag of the docker image to infer whether or not the used version
+     * of Localstack required services list.
+     */
+    private final boolean servicesListRequired;
```
Comment (eddumelendez) [inline modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java:71]:
```suggestion
    private final boolean servicesEnvVarRequired;
```

File: modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java [hunk 2]
```diff
@@ -93,11 +102,27 @@ public LocalStackContainer(final DockerImageName dockerImageName, boolean useLeg
         dockerImageName.assertCompatibleWith(DEFAULT_IMAGE_NAME);
 
         this.legacyMode = useLegacyMode;
+        this.servicesListRequired = isServicesListRequired(dockerImageName.getVersionPart());
 
         withFileSystemBind(DockerClientFactory.instance().getRemoteDockerUnixSocketPath(), "/var/run/docker.sock");
         waitingFor(Wait.forLogMessage(".*Ready\\.\n", 1));
     }
 
+    private static boolean isServicesListRequired(String version) {
```
Comment (eddumelendez) [inline modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java:111]:
```suggestion
    private static boolean isServicesEnvVarRequired(String version) {
```

File: modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java [hunk 3]
```diff
@@ -93,11 +102,27 @@ public LocalStackContainer(final DockerImageName dockerImageName, boolean useLeg
         dockerImageName.assertCompatibleWith(DEFAULT_IMAGE_NAME);
 
         this.legacyMode = useLegacyMode;
+        this.servicesListRequired = isServicesListRequired(dockerImageName.getVersionPart());
```
Comment (eddumelendez) [inline modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java:105]:
```suggestion
        this.servicesEnvVarRequired = isServicesEnvVarRequired(dockerImageName.getVersionPart());
```

File: modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java [hunk 4]
```diff
@@ -120,9 +145,13 @@ private static boolean shouldRunInLegacyMode(String version) {
     protected void configure() {
         super.configure();
 
-        Preconditions.check("services list must not be empty", !services.isEmpty());
+        if (servicesListRequired) {
```
Comment (eddumelendez) [inline modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java:148]:
```suggestion
        if (this.servicesEnvVarRequired) {
```

File: modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java [hunk 5]
```diff
@@ -120,9 +145,13 @@ private static boolean shouldRunInLegacyMode(String version) {
...
+            Preconditions.check("services list must not be empty", !services.isEmpty());
+        }
 
-        withEnv("SERVICES", services.stream().map(EnabledService::getName).collect(Collectors.joining(",")));
+        if (!services.isEmpty()) {
+            withEnv("SERVICES", services.stream().map(EnabledService::getName).collect(Collectors.joining(",")));
+        }
```
Comment (eddumelendez) [inline modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java:154]: > Scripts that depend on the output of /health being running will require setting setting EAGER_SERVICE_LOADING=1, which will start all services eagerly. This works together with the SERVICES variable to create the behavior from localstack <= 0.13.0.

Following the note from the release notes and do not break things


```suggestion
            if (this.servicesEnvVarRequired) {
                withEnv("EAGER_SERVICE_LOADING", "1");
            }
        }
```

File: modules/localstack/src/test/java/org/testcontainers/containers/localstack/LocalstackContainerTest.java [hunk 6]
```diff
@@ -347,4 +347,35 @@ public void s3EndpointHasProperRegion() {
...
+        @Test
+        public void servicesUseSingleEndpointUri() {
+            assertThat(localstack.getExposedPorts()).as("A single port is exposed").hasSize(1);
+            assertThat(localstack.getEndpointOverride(Service.SQS).toString())
+                .as("Endpoints for all services are the same")
+                .isEqualTo(localstack.getEndpointOverride(Service.S3).toString());
+        }
+
+        @Test
+        public void startsServicesLazily() {
```
Comment (eddumelendez) [inline modules/localstack/src/test/java/org/testcontainers/containers/localstack/LocalstackContainerTest.java:367]:
```suggestion
        public void s3ServiceStartLazily() {
```

File: modules/localstack/src/test/java/org/testcontainers/containers/localstack/LocalstackContainerTest.java [hunk 7]
```diff
@@ -347,4 +347,35 @@ public void s3EndpointHasProperRegion() {
...
+        );
+
+        @Test
+        public void servicesUseSingleEndpointUri() {
+            assertThat(localstack.getExposedPorts()).as("A single port is exposed").hasSize(1);
+            assertThat(localstack.getEndpointOverride(Service.SQS).toString())
+                .as("Endpoints for all services are the same")
+                .isEqualTo(localstack.getEndpointOverride(Service.S3).toString());
+        }
+
```
Comment (eddumelendez) [inline modules/localstack/src/test/java/org/testcontainers/containers/localstack/LocalstackContainerTest.java:365]: It's already covered https://github.com/testcontainers/testcontainers-java/blob/main/modules/localstack/src/test/java/org/testcontainers/containers/localstack/LegacyModeTest.java#L48

```suggestion
```

**Review by eddumelendez — CHANGES_REQUESTED** [review]
Thanks for your contribution, @maciejwalkowiak ! I have added some comments.

**Comment by eddumelendez** [comment]
Thanks for your contribution, @maciejwalkowiak ! I've applied the suggestions.

**Comment by maciejwalkowiak** [comment]
Thanks @eddumelendez!

**Pull request merged**